### DISCUSSION TITLE
COST OF USING ALTER TAG AND A VERTEX HAVE MULTI TAGS

### DISCUSSION THREAD

**Discussion by PhungQuocLuong-git** [vesoft-inc/nebula#5365] [Q&A] [answered]
Hello, 
I am curious about the flexibility of graph database when schema changes due to change of business.
Example, I have a tag  PRODUCT ( name string, price int ) and I have 1 million vertices for PRODUCT, because I have the need to manage the color of these product. 
I wonder if I add a tag PRODUCT_COLOR then apply it for all product (1 vertex will have 2 tag : PRODUCT and PRODUCT_COLOR )  will it be more effective than I use ALTER TAG to add a property to tag  PRODUCT ( tag PRODUCT  will have 3 prop : name, price and color)
By the way, I want to ask 2 more questions:
1) In case I add a tag when the business changes, 1 vertex will have many tags over time. Is this the flexible of graph database or It will the bad case of modeling ( bad performance ) ?
2) In case I use ALTER TAG, does Nebula graph add color property to 1 million vertices ?  ( I consider about performance when using ALTER TAG for a large of data ).
Thanks :))

**Comment by wey-gu** [comment, marked as answer]
Great question.

Generally
- one tag per vertex is one record(underlying rocksdb key-value).
- alter tag won't directly write alter-ed props, unless we write a certain record(vertex-tag record), but before it's written, we could read based on an alter-ed schema, with new default values.

So back to the questions:

- In case I add a tag when the business changes, 1 vertex will have many tags over time. Is this the flexible of graph database or It will the bad case of modeling ( bad performance ) ?

> wey: it will be ok, but when we are doing query, it's always more performant to provide explicit information, i.e. ` RETURN v. PRODUCT.foo, v.PRODUCT_COLOR.bar ` is better than `RETURN v`

- In case I use ALTER TAG, does Nebula graph add color property to 1 million vertices ? ( I consider about performance when using ALTER TAG for a large of data ).

> wey: no, unless you do update/insert to change the color prop from the default color value, otherwise it left vertices(in PRODUCT tag) unchanged.

ref: https://www.nebula-graph.io/tags/architecture
